sciencebeam/pipeline_runners/beam_pipeline_runner.py

```diff
@@ -0,0 +1,259 @@
+from __future__ import absolute_import
+
+import argparse
+import os
+import logging
+import mimetypes
+
+import apache_beam as beam
+from apache_beam.options.pipeline_options import PipelineOptions, SetupOptions
+from apache_beam.metrics.metric import Metrics
+
+from sciencebeam_gym.utils.collection import (
+  extend_dict
+)
+
+from sciencebeam_gym.beam_utils.utils import (
+  TransformAndCount,
+  TransformAndLog,
+  MapOrLog,
+  PreventFusion
+)
+
+from sciencebeam_gym.beam_utils.files import (
+  ReadFileList,
+  FindFiles
+)
+
+from sciencebeam_gym.beam_utils.io import (
+  read_all_from_path,
+  save_file_content
+)
+
+from sciencebeam_gym.beam_utils.main import (
+  add_cloud_args,
+  process_cloud_args,
+  process_sciencebeam_gym_dep_args
+)
+
+from sciencebeam_gym.preprocess.preprocessing_utils import (
+  join_if_relative_path,
+  get_output_file
+)
+
+from sciencebeam.config.app_config import get_app_config
+
+from .simple_pipeline_runner import _pipeline
+
+LOGGER = logging.getLogger(__name__)
+
+def get_logger():
+  return logging.getLogger(__name__)
+
+class MetricCounters(object):
+  FILES = 'files'
+
+class DataProps(object):
+  SOURCE_FILENAME = 'source_filename'
+  FILENAME = 'filename'
+  CONTENT = 'content'
+  TYPE = 'type'
+
+def FileUrlSource(opt):
+  if opt.source_file_list:
+    return ReadFileList(opt.source_file_list, column=opt.source_file_column, limit=opt.limit)
+  else:
+    return FindFiles(join_if_relative_path(opt.base_data_path, opt.source_path))
+
+def ReadFileContent():
+  return "ReadFileContent" >> TransformAndCount(
+    beam.Map(lambda file_url: {
+      DataProps.SOURCE_FILENAME: file_url,
+      DataProps.FILENAME: file_url,
+      DataProps.CONTENT: read_all_from_path(file_url)
+    }),
+    MetricCounters.FILES
+  )
+
+def get_step_error_counter(step):
+  return 'error_%s' % step
+
+def get_step_ignored_counter(step):
+  return 'ignored_%s' % step
+
+def get_step_processed_counter(step):
+  return 'processed_%s' % step
+
+def execute_or_skip_step(step):
+  supported_types = step.get_supported_types()
+  processed_counter = Metrics.counter('PipelineStep', get_step_processed_counter(step))
+  ignored_counter = Metrics.counter('PipelineStep', get_step_ignored_counter(step))
+
+  def wrapper(x):
+    data_type = x['type']
+    if data_type in supported_types:
+      get_logger().debug('excuting step %s: %s (%s)', step, x.keys(), data_type)
```

Comment:
This function will actually be serialised and deserialised again on a worker. I'm not sure whether it's good to serialise the logger in that case.